### PR TITLE
Prevent carrier negative range

### LINKED ISSUE #37154: Feature flag Carrier page - Exception displayed when creating a carrier with negative ranges
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

When I create a carrier with overlapping ranges, I have an exception instead of just a warning.

Screenshot 2024-10-15 at 15 23 45

https://github.com/user-attachments/assets/0359bf8b-c04f-410e-8a4e-aa52035fff9a

### Expected behavior

The exception should not be displayed.

### Steps to reproduce

1. Go to BO > Shipping > Carriers
2. Edit any carrier > Go to Shipping locations and costs
3. Click on manage ranges
4. Put any negative number on the first range for the Minimum amount > Apply
5. Save > See exception `Carrier ranges are overlapping` :x:

### PrestaShop version(s) where the bug happened

9.0.0

### PHP version(s) where the bug happened

8.1

### If your bug is related to a module, specify its name and its version

_No response_

### Your company or customer's name goes here (if applicable).

_No response_

### PR DESCRIPTION
**Pull request by tleon**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Prevent the user to add a range with a negative number or an empty field.
| Type?             | bug fix 
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | Describe in the associated issue
| UI Tests          | https://github.com/tleon/ga.tests.ui.pr/actions/runs/11840524867
| Fixed issue or discussion?     | #37154
| Sponsor company   | PrestaShop SA

### CODE REVIEW

**Comment by ps-jarvis** [comment]
Hi, thanks for this contribution!

I found some issues with the Pull Request description:

* The `type` should be one of these: `new feature`, `improvement`, `bug fix` or `refacto`. ([Read explanation](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#branch))

Would you mind having a look at it? This will help us understand how interesting your contribution is, thank you very much!

(Note: this is an automated message, but answering it will reach a real human)

File: src/Core/Domain/Carrier/ValueObject/CarrierRangeZone.php [hunk 1]
```diff
@@ -109,6 +109,12 @@ private function assertRanges(array $ranges)
 
         // Then, we can check if ranges are overlapping or not
         foreach ($ranges as $range) {
+            if ($range['range_from'] < 0) {
```
Comment (jolelievre) [inline src/Core/Domain/Carrier/ValueObject/CarrierRangeZone.php]: You should check range_to as well no?

File: tests/Integration/Behaviour/Features/Scenario/Carrier/Ranges/carrier_ranges.feature [hunk 2]
```diff
@@ -154,3 +154,13 @@ Feature: Carrier ranges
       | shippingMethod   | weight                             |
       | zones            |                                    |
     Then carrier should throw an error with error code "INVALID_ZONE_MISSING"
+
+  Scenario: Adding negative ranges in carrier
+    When I create carrier "carrier1" with specified properties:
+      | name             | Carrier 1                          |
+      | shippingMethod   | weight                             |
+      | zones            | zone1, zone2                       |
+    Then I set ranges for carrier "carrier1" with specified properties for all shops:
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Carrier/Ranges/carrier_ranges.feature:163]: If a rule is added on the ranges you should add another error check with range_to negative

**Review by jolelievre — COMMENTED** [review]
Thanks @tleon 

I think we could add one last constraint in the Symfony form type here:
https://github.com/PrestaShop/PrestaShop/blob/68b96d005e02ebebf05cae6f518a17a02a2a313a/src/PrestaShopBundle/Form/Admin/Improve/Shipping/Carrier/Type/CostsRangeType.php#L49-L50

Both from and to fields should have a constraint that forces minimum value to 0

**Comment by ps-jarvis** [comment]
This pull request seems to contain new translation strings. I have summarized them below to ease up review:

- `Admin.Shipping.Feature`
    - [ ] `Maximum`
    - [ ] `Action`
    - [ ] `You can not add a negative range`
    - [ ] `Range`
    - [ ] `Price (VAT excl.)`

(Note: this is an automated message, but answering it will reach a real human)

File: src/PrestaShopBundle/Form/Admin/Improve/Shipping/Carrier/Type/CostsRangeType.php [hunk 3]
```diff
@@ -46,13 +47,24 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
+            ->add('from', HiddenType::class, [
+                'constrains' => [
+                    new PositiveOrZero([
+                        'message' => 'The value must be a positive number.',
+                    ]),
+                ],
+            ])
+            ->add('to', HiddenType::class, [
+                'constrains' => [
+                    new PositiveOrZero([
```
Comment (jolelievre) [inline src/PrestaShopBundle/Form/Admin/Improve/Shipping/Carrier/Type/CostsRangeType.php:59]: I wonder if this one should be strictly positive, but that's just a suggestion/question I'm asking myself I guess QA or user feedback will let us know

**Review by jolelievre — APPROVED** [review]
Thanks @tleon

**Review by nicosomb — DISMISSED** [review]

**Review by boherm — DISMISSED** [review]

**Review by florine2623 — COMMENTED** [review]
Hello @tleon ,

Following the steps to reproduce in the issue, I have a big exception at step 2.

Screenshot 2024-11-05 at 15 54 49

```
HTTP 500 Internal Server Error
An error has occurred resolving the options of the form "Symfony\Component\Form\Extension\Core\Type\HiddenType": The option "constrains" does not exist. Defined options are: "action", "alert_message", "alert_position", "alert_title", "alert_type", "allow_extra_fields", "allow_file_upload", "attr", "attr_translation_parameters", "auto_initialize", "block_name", "block_prefix", "by_reference", "column_breaker", "columns_number", "compound", "constraints", "csrf_field_name", "csrf_message", "csrf_protection", "csrf_token_id", "csrf_token_manager", "data", "data_class", "default_empty_data", "disabled", "disabled_value", "disabling_switch", "disabling_switch_event", "empty_data", "empty_view_data", "error_bubbling", "error_mapping", "external_link", "extra_fields_message", "form_attr", "form_theme", "getter", "help", "help_attr", "help_html", "help_translation_parameters", "hint", "inherit_data", "invalid_message", "invalid_message_parameters", "is_empty_callback", "label", "label_attr", "label_format", "label_help_box", "label_html", "label_subtitle", "label_tab", "label_tag_name", "label_translation_parameters", "mapped", "method", "modify_all_shops", "multistore_configuration_key", "multistore_dropdown", "post_max_size_message", "priority", "property_path", "required", "row_attr", "setter", "switch_state_on_disable", "translation_domain", "trim", "upload_max_size_message", "use_default_themes", "validation_groups".
```

Am I missing something ?
Thanks!

File: src/PrestaShopBundle/Form/Admin/Improve/Shipping/Carrier/Type/CostsRangeType.php [hunk 4]
```diff
@@ -46,13 +47,24 @@ public function buildForm(FormBuilderInterface $builder, array $options)
             ->add('range', TextPreviewType::class, [
                 'label' => $this->trans('Range', 'Admin.Shipping.Feature'),
             ])
-            ->add('from', HiddenType::class)
-            ->add('to', HiddenType::class)
+            ->add('from', HiddenType::class, [
+                'constrains' => [
```
Comment (jolelievre) [inline src/PrestaShopBundle/Form/Admin/Improve/Shipping/Carrier/Type/CostsRangeType.php]:
```suggestion
                'constraints' => [
```

File: src/PrestaShopBundle/Form/Admin/Improve/Shipping/Carrier/Type/CostsRangeType.php [hunk 5]
```diff
@@ -46,13 +47,24 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
+                    new PositiveOrZero([
+                        'message' => 'The value must be a positive number.',
+                    ]),
+                ],
+            ])
+            ->add('to', HiddenType::class, [
+                'constrains' => [
```
Comment (jolelievre) [inline src/PrestaShopBundle/Form/Admin/Improve/Shipping/Carrier/Type/CostsRangeType.php]:
```suggestion
                'constraints' => [
```

Comment (jolelievre) [reply src/PrestaShopBundle/Form/Admin/Improve/Shipping/Carrier/Type/CostsRangeType.php]: This should fix the bug detected by QA

**Review by jolelievre — CHANGES_REQUESTED** [review]

**Comment by tleon** [comment]
@florine2623 should be fixed now. Tests are running

**Review by florine2623 — COMMENTED** [review]
Hello @tleon ,

No more exception is displayed. I can't Save a range with a minus sign ✅ 

Although, I don't have any clear error message. Maybe a message can be added in the general warning and displayed when I enter the incorrect data ?

Screenshot 2024-11-12 at 17 19 00

I have relaunch the automated test with the latest modifications of the PR : 
https://github.com/florine2623/testing_pr/actions/runs/11799652362 ✅ 

Thanks!

**Comment by tleon** [comment]
Added an error message when trying to submit a negative range 
image
Can change wording if it's not ok like this.

**Review by jolelievre — APPROVED** [review]

**Review by florine2623 — APPROVED** [review]
Thanks ! The error message error is well displayed when a negative range is typed.

It is QA ✅

**Pull request merged**